lib/SPIRV/SPIRVTypeScavenger.cpp

```diff
@@ -53,156 +137,497 @@
         correctUseTypes(I);
       }
     }
   }
 
-  // Go through all of the types we have collected, and if any are still
-  // deferred, assign them a fallback i8* type.
+  // If there are any type variables we couldn't resolve, fallback to assigning
+  // them as an i8* type.
   Type *Int8Ty = Type::getInt8Ty(M.getContext());
-  for (const auto &Pair : DeducedTypes) {
-    if (auto *Untyped = dyn_cast<DeferredType *>(Pair.second)) {
-      LLVM_DEBUG(dbgs() << "No inferrable type for " << *Pair.first << "\n");
-      fixType(*Untyped, Int8Ty);
-      DeducedTypes[Pair.first] = Int8Ty;
+  for (auto [TypeVarNum, TypeVar] : enumerate(TypeVariables)) {
+    unsigned PrimaryVar = UnifiedTypeVars.join(TypeVarNum, TypeVarNum);
+    Type *LeaderTy = TypeVariables[PrimaryVar];
+    if (TypeVar)
+      TypeVar = substituteTypeVariables(TypeVar);
+    if (LeaderTy)
+      LeaderTy = substituteTypeVariables(LeaderTy);
+    assert((!TypeVar || LeaderTy == TypeVar) &&
+           "Inconsistent type variable unification");
+    if (!TypeVar) {
+      TypeVar = LeaderTy ? LeaderTy : Int8Ty;
     }
+    TypeVariables[TypeVarNum] = TypeVar;
+    LLVM_DEBUG(dbgs() << "Type variable " << TypeVarNum << " resolved to "
+                      << *TypeVar << "\n");
   }
   return;
 }
 
-static Type *getPointerUseType(Function *F, Op Opcode, unsigned ArgNo) {
-  switch (Opcode) {
-  case OpAtomicLoad:
-  case OpAtomicExchange:
-  case OpAtomicCompareExchange:
-  case OpAtomicIAdd:
-  case OpAtomicISub:
-  case OpAtomicFAddEXT:
-  case OpAtomicSMin:
-  case OpAtomicUMin:
-  case OpAtomicFMinEXT:
-  case OpAtomicSMax:
-  case OpAtomicUMax:
-  case OpAtomicFMaxEXT:
-  case OpAtomicAnd:
-  case OpAtomicOr:
-  case OpAtomicXor:
-    if (ArgNo == 0)
-      return F->getReturnType();
-    return nullptr;
-  case OpAtomicStore:
-    if (ArgNo == 0)
-      return F->getArg(3)->getType();
-    return nullptr;
-  default:
-    return nullptr;
-  }
-}
-
 bool SPIRVTypeScavenger::typeIntrinsicCall(
-    CallBase &CB, SmallVectorImpl<std::pair<unsigned, DeducedType>> &ArgTys) {
+    CallBase &CB, SmallVectorImpl<TypeRule> &TypeRules) {
   Function *TargetFn = CB.getCalledFunction();
   assert(TargetFn && TargetFn->isDeclaration() &&
          "Call is not an intrinsic function call");
   LLVMContext &Ctx = TargetFn->getContext();
 
+  StringRef DemangledName;
+  if (oclIsBuiltin(TargetFn->getName(), DemangledName) ||
+      isDecoratedSPIRVFunc(TargetFn, DemangledName)) {
+    Op OC = getSPIRVFuncOC(DemangledName);
+    switch (OC) {
+    case OpAtomicLoad:
+    case OpAtomicExchange:
+    case OpAtomicCompareExchange:
+    case OpAtomicIAdd:
+    case OpAtomicISub:
+    case OpAtomicFAddEXT:
+    case OpAtomicSMin:
+    case OpAtomicUMin:
+    case OpAtomicFMinEXT:
+    case OpAtomicSMax:
+    case OpAtomicUMax:
+    case OpAtomicFMaxEXT:
+    case OpAtomicAnd:
+    case OpAtomicOr:
+    case OpAtomicXor:
+      TypeRules.push_back(TypeRule::pointsTo(CB, 0, CB.getType()));
```

Comment:
nit here and after for TypeRule::pointsTo because I'm not sure if it's 'easy' to copy PointerUnion without overhead (please share if you have another data)
```suggestion
      TypeRules.emplace_back(TypeRule::pointsTo(CB, 0, CB.getType()));
```